packages/fx-core/src/component/generator/utils.ts

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import Mustache from "mustache";
+import path from "path";
+import * as fs from "fs-extra";
+import { selectTag } from "../../common/template-utils/templates";
+import { fetchTemplateTagList } from "../../common/template-utils/templatesUtils";
+import {
+  defaultTimeoutInMs,
+  defaultTryLimits,
+  templateFileExt,
+  templateTagListUrl,
+} from "./constant";
+import {
+  FetchSampleUrlWithTagError,
+  FetchZipFromUrlError,
+  TemplateZipFallbackError,
+  UnzipError,
+} from "./error";
+import { GenerateAction, GenerateActionName } from "./generateAction";
+import { GenerateContext } from "./generateContext";
+import AdmZip from "adm-zip";
+import { SampleInfo, sampleProvider } from "./sample";
+
+export async function fetchZipUrl(
+  name: string,
+  baseUrl: string,
+  tryLimits = defaultTryLimits,
+  timeoutInMs = defaultTimeoutInMs
+): Promise<string> {
+  const tags = await fetchTemplateTagList(templateTagListUrl, tryLimits, timeoutInMs);
+  const selectedTag = selectTag(tags.replace(/\r/g, "").split("\n"));
+  if (!selectedTag) {
+    throw new Error(`Failed to find valid template for ${name}`);
+  }
+  return `${baseUrl}/${selectTag}/${templateZipName(name)}`;
+}
+
+export const templateZipName = (templateName: string): string => `${templateName}.zip`;
+
+export function renderTemplateFileData(
+  fileName: string,
+  fileData: Buffer,
+  variables?: { [key: string]: string }
+): string | Buffer {
+  //only mustache files with name ending with .tql
+  if (path.extname(fileName) === templateFileExt) {
+    return Mustache.render(fileData.toString(), variables);
+  }
+  // Return Buffer instead of string if the file is not a template. Because `toString()` may break binary resources, like png files.
+  return fileData;
+}
+
+export function renderTemplateFileName(
+  fileName: string,
+  fileData: Buffer,
+  variables?: { [key: string]: string }
+): string {
+  return Mustache.render(fileName, variables);
+}
+
+export function genFileDataRenderReplaceFn(variables: { [key: string]: string }) {
```

Comment:
In this class, why do we need to have some unnecessary wrapper like this. I'm a little confused on so many utils class, can you show me more high level about the relationships between the core Generator logic and those kinds of utils?